blog/2020-07-13-alpine-python.md

```diff
@@ -0,0 +1,156 @@
+---
+author: Pascal Vizeli
+authorURL: https://github.com/pvizeli
+authorTwitter: pvizeli
+title: "Alpine with musl and Python"
+---
+
+It is well known that we use Alpine for most of our Containers. It is the perfect distribution for containers because it is small (BusyBox based), available for a lot of CPU architectures, and the package system is slim. Alpine and musl are new and had at least a significant development progress. That means many statements are today just wrong. I will clean up this incorrect information based on my research to provide the fastest way to run Home Assistant. This blogpost is not a musl vs. glibc flame ware; we use glibc on our OS. I will clean up old or half-truth, which is too long now in people's minds.
+
+For tests, I used the images from [Docker Python library](https://github.com/docker-library/python), and the result is published to our [base images](https://github.com/home-assistant/docker-base). I used pyperformance for labs testing and the Home Assistant internal benchmark tools for more Real-life information. The test environment is running inside a container on the same docker host.
+
+## C/POSIX standard library
+
+I read often: Python is slower if they use musl as default C library. This fact is not 100% correct. If both Python runtime was compiled with the same GCC and with `-O3`, the glibc variant is a bit faster in the lab benchmark, but in the real world, it makes not a difference. Alpine compile it for `-Os` and most distro with `-O2`, which gives the often written difference between the Python runtime interpreters, but with the same compiler optimizing, musl based Python runtimes have no negative side effect.
+
+But there is a game-changer which makes the musl one more useful as the glibc based runtime. It is the memory allocator [jemalloc](http://jemalloc.net/), a general-purpose malloc implementation that emphasizes fragmentation avoidance and scalable concurrency support. There is an interesting effect which I found on some blogpost about Rust. There were some developers who saw that musl is much faster as glibc with jemalloc, but glibc is slower with jemalloc as without. For sure, the benefit with glibc and jemalloc is not the speed as they optimize memory management, but musl get both benefits. While the difference between pure musl and glibc can be ignored, the difference between musl + jemalloc and glibc are substantial (with disabled GCC memory allocator built-in optimization). Yes, jemalloc is compatible today with musl, and there was a time which it was not.
```

Comment:
```suggestion
But there is a game-changer, which makes the musl one more useful compared to the glibc based runtime. It is the memory allocator [jemalloc](http://jemalloc.net/), a general-purpose malloc implementation that emphasizes fragmentation avoidance and scalable concurrency support. There is an interesting effect, which I found on some blogpost about Rust. There were some developers who saw that musl is much faster as glibc with jemalloc, but glibc is slower with jemalloc as without. For sure, the benefit with glibc and jemalloc is not the speed as they optimize memory management, but musl get both benefits. While the difference between pure musl and glibc can be ignored, the difference between musl + jemalloc and glibc are substantial (with disabled GCC memory allocator built-in optimization). Yes, today's jemalloc is compatible with musl (there was a time which it was not).
```

Reply:
> There were some developers who saw that musl is much faster as glibc with jemalloc, but glibc is slower with jemalloc as without.

IMO this could benefit from re-phrasing. Frankly, I've read it quite a few times but still don't quite understand what "setups" are being compared here. And it could be just me...

for example `that musl is much faster as glibc` if it is faster as in "comparison" then we should probably use "than" as in `musl is much faster than glibc` or if you meant they offer similar speeds then it should be `that musl is as fast as glibc`

> but glibc is slower with jemalloc as without

Same here. is it just me having issues parsing it? (wouldn't be the 1st time :)

Reply:
> but glibc is slower with jemalloc as without

but glibc with jemalloc is slower than glibc without it.

Reply:
so as I understand, we are comparing:
1. musl with jemalloc
2. glibc with jemalloc
3. glibc without jemalloc
so we just should rank the above?